### ISSUE TITLE
Documentation

### ISSUE THREAD

**Issue by zoq** [mlpack/ensmallen#25]
I think right now we can keep the documentation relatively simple, I think something like this could work?

### AdaDelta
Adadelta is an extension of Adagrad that adapts learning rates based on a moving window of gradient updates, instead of accumulating all past gradients. Instead of accumulating all past squared gradients, the sum of gradients is recursively defined as a decaying average of all past squared gradients.

**AdaDelta(stepSize, batchSize, rho, epsilon , maxIterations, tolerance, shuffle);**
**AdaDelta(stepSize, batchSize);**
**AdaDelta(stepSize);**

#### Input Options:
<table>
<colgroup>
<col width="30%" />
<col width="70%" />
</colgroup>
<thead>
<tr class="header">
</tr>
</thead>
<tbody>
<tr>
<td markdown="span">**stepSize** (double)</td>
<td markdown="span">Step size for each iteration.</td>
</tr>
<tr>
<td markdown="span">**batchSize** (int)</td>
<td markdown="span">Number of points to process in one step.</td>
</tr>
<tr>
<td markdown="span">**rho** (double)</td>
<td markdown="span">Smoothing constant. Corresponding to fraction of gradient to keep at each time step.</td>
</tr>
<tr>
<td markdown="span">**epsilon** (double)</td>
<td markdown="span">Value used to initialise the mean squared gradient parameter.</td>
</tr>
<tr>
<td markdown="span">**maxIterations** (int)</td>
<td markdown="span">Maximum number of iterations allowed (0 means no limit).</td>
</tr>
<tr>
<td markdown="span">**tolerance** (double)</td>
<td markdown="span">Maximum absolute tolerance to terminate algorithm.</td>
</tr>
<tr>
<td markdown="span">**shuffle** (bool)</td>
<td markdown="span">If true, the function order is shuffled; otherwise, each function is visited in linear order.</td>
</tr>
</tbody>
</table>

#### Example:

```c++
 AdaDelta optimizer(1.0, 1, 0.99, 1e-8, 5000000, 1e-9, true)
 RosenbrockFunction f;
 arma::mat coordinates = f.GetInitialPoint();
 optimizer.Optimize(f, coordinates);
````

### See also:

* Adadelta - an adaptive learning rate method
* Adagrad

**Comment by rcurtin** [comment]
That looks great to me.  I started sketching some simple documentation for the different `FunctionType` abstractions so if you want to handle the optimizers I can handle that part.

**Comment by conradsnicta** [comment]
Looks good.  For each optimizer algorithm I recommend to add a link (in the See also part) to a corresponding Wikipedia article, and/or arXiv paper, and/or DOI linked paper.

Otherwise someone who is not versed with all the variants (or even the general nature of each algorithm) will have no real idea which algorithm would be the most suitable for their particular problem.

Furthermore, having links to reference material increases the professional appearance and brings about reassurance that ensmallen didn't just come up with half these algorithms out of thin air.

Example in Armadillo: http://arma.sourceforge.net/docs.html#polyfit

**Comment by zoq** [comment]
Not sure if we like to go with the current style (highly based on the simple/clean/great armadillo doc style) and still some optimizer and details left: http://kurg.org/misc/ensmallen/docs.html

**Comment by conradsnicta** [comment]
Looks mostly ok.  A few suggestions:
* At the top of the page, change "Optimizer" to "Optimizers"
* For each optimizer section (eg. AdaDelta), use the horizontal line only for separating the sections from each other. Right now the second horizontal line (used as an underline of sorts) breaks the flow.
* The attributes look a bit confusing.  At first look **stepSize**(double) looks like a function, which is definitely not the intended effect. This can throw off beginners and casual users.  Maybe change this to something along these lines:
   * stepSize  &nbsp;&nbsp;&nbsp; Step size for each iteration (type: double)

* It's not clear what is meant by "For convenience the optimizer attributes can be modified through setters and getters which share names with their respective data members".  Some people may work this out through trial and error, but most people are lazy and easily irritated if it's not immediately clear how to use something.  The explanation should be clarified along these lines:
    * The optimizer attributes can be modified through setters and getters which share names with their respective data members. As an example, for an instance of AdaDelta named as _optimizer_, the attribute **stepSize** can be obtained and modified using the following form:
        * double s = optimizer.get_stepSize();
        * optimizer.set_stepSize(s);

* If the above explanation applies to all the optimizers, then it should be a separate standalone section, instead of being repeated for each optimizer.

* I recommend to have a separate "Example Program" section which includes a simple and complete example program that shows how to use an optimizer.  Copying the example out of the paper should be sufficient.

**Comment by rcurtin** [comment]
Nice!  Looks good with a quick skim (I'll read through in full sometime in the future).  I'm not picky on the style, maybe we can just match it to whatever is set up for the website.  About the types, maybe it's better to write, e.g., "double **stepSize**"?

**Comment by conradsnicta** [comment]
That would work as well.  Maybe have the types in a separate column, for a clean look.

**Comment by zoq** [comment]
Thanks for the comments, the latest version will be available  here: http://kurg.org/misc/ensmallen/docs.html

**Comment by conradsnicta** [comment]
Looks good. Minor issues:

* In the last section titled Stochastic Gradient Descent with Restarts (SGDR), there is an errant bullet point near the start of the section
* In the Katyusha section, there's a bit of line noise: "... on top of Nesterovâ€™s momentum."
* In the L-BFGS section, also line noise: "Broydenâ€“Fletcherâ€“Goldfarbâ€“Shanno"

**Comment by conradsnicta** [comment]
I've committed the initial website sources directly to https://github.com/mlpack/ensmallen.org
A few points to note:
* The download link on the home page is currently simply set to "ensmallen-1.10.0.tar.xz" within the same directory; this file obviously doesn't exist yet
* I've embedded the current draft of the documentation into docs.html; there is html code for the layout/menus/etc above and below the embedded documentation; there is also proposed additional content at the bottom of the page

* The embedded draft of the documentation has a bunch of html errors. You can use "weblint" to detect them.  On Fedora the weblint program is in the perl-HTML-Lint-2.30 package.

* The pages will rescale automatically between desktop and mobile phone views.  You can simulate a mobile phone view by horizontally shrinking your browser window; once the window is small enough, the layout will automatically change

* The logo shrinks/expands as the browser windows shrinks/expands (up to a limit)

* The logo currently points to http://ensmallen.org

**Comment by conradsnicta** [comment]
Ryan has reworked the documentation, so we can close this.
